pkg/noderesourcetopology/filter.go

```diff
@@ -0,0 +1,254 @@
+}
+
+type NUMANodeList []NUMANode
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (tm *TopologyMatch) Name() string {
+	return Name
+}
+
+func extractResources(zone topologyv1alpha1.Zone) v1.ResourceList {
+	res := make(v1.ResourceList)
+	for _, resInfo := range zone.Resources {
+		quantity, err := resource.ParseQuantity(resInfo.Allocatable.String())
+		if err != nil {
+			klog.Errorf("Failed to parse %s", resInfo.Allocatable.String())
+			continue
+		}
+		res[v1.ResourceName(resInfo.Name)] = quantity
+	}
+	return res
+}
+
+func SingleNUMAContainerLevelHandler(pod *v1.Pod, zones topologyv1alpha1.ZoneList) *framework.Status {
+	klog.V(5).Infof("Single NUMA node handler")
+
+	// prepare NUMANodes list from zoneMap
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		resBitmask := checkResourcesForNUMANodes(nodes, container.Resources.Requests, qos)
+		if resBitmask.IsEmpty() {
+			// definitely we can't align container, so we can't align a pod
+			return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Cannot align container: %s", container.Name))
+		}
+	}
+	return nil
+}
+
+func checkResourcesForNUMANodes(nodes NUMANodeList, resources v1.ResourceList, qos v1.PodQOSClass) bm.BitMask {
+	bitmask := bm.NewEmptyBitMask()
+	bitmask.Fill()
+
+	zeroQuantity := resource.MustParse("0")
+	for resource, quantity := range resources {
+		resourceBitmask := bm.NewEmptyBitMask()
+		for _, numaNode := range nodes {
+			numaQuantity, ok := numaNode.Resources[resource]
+			// if can't find requested resource on the node - skip (don't set it as available NUMA node)
+			// if unfound resource has 0 quantity probably this numa node can be considered
+			if !ok && quantity.Cmp(zeroQuantity) != 0 {
+				continue
+			}
+			// Check for the following:
+			// 1. set numa node as possible node if resource is memory or Hugepages (until memory manager will not be merged and
+			// memory will not be provided in CRD
+			// 2. set numa node as possible node if resource is cpu and it's not guaranteed QoS, since cpu will flow
+			// 3. set numa node as possible node if zero quantity for non existing resource was requested (TODO check topology manaager behaviour)
+			// 4. otherwise check amount of resources
+			if resource == v1.ResourceMemory ||
+				strings.HasPrefix(string(resource), string(v1.ResourceHugePagesPrefix)) ||
+				resource == v1.ResourceCPU && qos != v1.PodQOSGuaranteed ||
+				quantity.Cmp(zeroQuantity) == 0 ||
+				numaQuantity.Cmp(quantity) >= 0 {
+				resourceBitmask.Add(numaNode.NUMAID)
+			}
+		}
+		bitmask.And(resourceBitmask)
+	}
+	return bitmask
+}
+
+func SingleNUMAPodLevelHandler(pod *v1.Pod, zones topologyv1alpha1.ZoneList) *framework.Status {
+	klog.V(5).Infof("Pod Level Resource handler")
+	resources := make(v1.ResourceList)
+
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
```

Comment:
For PodLevel handling, I'm not quite sure it's correct to simply aggregate InitContainers and InitContainers. The relation in Kubernetes Resource Model to evaluate a pod's request is:

```
resourceRequest = max(sum(podSpec.Containers), podSpec.InitContainers) + overHead
```

So probably we should adapt to KSM accordingly.

Reply:
probably yes, but current implementation of TM
https://github.com/kubernetes/kubernetes/blob/9657dd77a3563ffc7dbd061d51f34f714b1f8d5c/pkg/kubelet/cm/topologymanager/scope_pod.go#L57
takes it into account, for both container level and pod level. It's hard for me to image where it could be useful, I mean NUMA/CPU affinity for InitContainer. We'll discuss it with TM authors.

Reply:
SG, let's keep it consistent with TM for now. Good to leave a comment here.